### PR TITLE
Adds Spot market request import

### PR DESCRIPTION
**Pull request by displague**
This is the last resource to implement importers from the list originally prepared at https://github.com/packethost/terraform-provider-packet/issues/51.

I've also added import documentation for some resources that supported it but did not document it.  (I did not attempt to document the resources with composite keys).

### CODE REVIEW

**Comment by t0mk** [comment]
@displague 

I run the test and the import seems not to work [0]. The TF resource is not populated properly. Please add a setMap in resourceMetalSpotMarketRequestRead.

Params of SMR are https://github.com/packethost/packngo/blob/master/spotmarketrequest.go#L28

[0]
```
Cmetro
    testing.go:569: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) {
        }
        
        
        (map[string]string) (len=16) {
         (string) (len=11) "devices_max": (string) (len=1) "1",
         (string) (len=11) "devices_min": (string) (len=1) "1",
         (string) (len=21) "instance_parameters.#": (string) (len=1) "1",
         (string) (len=32) "instance_parameters.0.always_pxe": (string) (len=5) "false",
         (string) (len=35) "instance_parameters.0.billing_cycle": (string) (len=6) "hourly",
         (string) (len=32) "instance_parameters.0.customdata": (string) "",
         (string) (len=33) "instance_parameters.0.description": (string) "",
         (string) (len=30) "instance_parameters.0.hostname": (string) (len=14) "tfacc-testspot",
         (string) (len=37) "instance_parameters.0.ipxe_script_url": (string) "",
         (string) (len=28) "instance_parameters.0.locked": (string) "",
         (string) (len=38) "instance_parameters.0.operating_system": (string) (len=12) "ubuntu_18_04",
         (string) (len=26) "instance_parameters.0.plan": (string) (len=11) "baremetal_0",
         (string) (len=39) "instance_parameters.0.termintation_time": (string) "",
         (string) (len=30) "instance_parameters.0.userdata": (string) "",
         (string) (len=13) "max_bid_price": (string) (len=5) "0.096",
         (string) (len=16) "wait_for_devices": (string) (len=4) "true"
        }
```

None of those fields is set

**Comment by t0mk** [comment]
@displague I think I can fix this today.

**Comment by t0mk** [comment]
I tried to read the instance_parameters back from the first device of the spot market request resource. It won't work 100%, for example of there's a variable name in hostname in the intance params for the SMR (e.g. sth like `host_$count`), it will not be the same in the hostname of the first created device.

Also, I still get nonempty diff after the import:
`         max_bid_price:                           "0.43" => "0.432" (forces new resource)`

This might be some rounding issue? I will check tomorrow.

**Comment by displague** [comment]
> max_bid_price: "0.43" => "0.432" (forces new resource)

If this isn't a rounding error, we might need to make this a computed field or IgnoreDiff based on some rounding or drift.  If "drift" is built into the resource we may want to expose this as a value the user can set for ignore purposes (maybe we set a default of ±0.01).  It's also possible that this is an API bug.  Looking forward to whatever you may find out.

**Comment by t0mk** [comment]
hmm, when I fix the drift, I get back to this error;

```
    testing.go:569: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) {
        }
        
        
        (map[string]string) (len=13) {
         (string) (len=21) "instance_parameters.#": (string) (len=1) "1",
         (string) (len=32) "instance_parameters.0.always_pxe": (string) (len=5) "false",
         (string) (len=35) "instance_parameters.0.billing_cycle": (string) (len=6) "hourly",
         (string) (len=32) "instance_parameters.0.customdata": (string) "",
         (string) (len=33) "instance_parameters.0.description": (string) "",
         (string) (len=30) "instance_parameters.0.hostname": (string) (len=14) "tfacc-testspot",
         (string) (len=37) "instance_parameters.0.ipxe_script_url": (string) "",
         (string) (len=28) "instance_parameters.0.locked": (string) "",
         (string) (len=38) "instance_parameters.0.operating_system": (string) (len=12) "ubuntu_18_04",
         (string) (len=26) "instance_parameters.0.plan": (string) (len=13) "c3.medium.x86",
         (string) (len=39) "instance_parameters.0.termintation_time": (string) "",
         (string) (len=30) "instance_parameters.0.userdata": (string) "",
         (string) (len=16) "wait_for_devices": (string) (len=4) "true"
        }

```

**Comment by displague** [comment]
Getting a little closer. One to go:
```
        Error: errors during refresh: instance_parameters.0.user_ssh_keys: '': source data must be an array or slice, got string
```

**Comment by displague** [comment]
Tests are failing because SpotMarketRequest does not include Parameters:
https://github.com/packethost/packngo/issues/277

I'll remove the setting of Parameters after creation to address this.

**Comment by displague** [comment]
Failing test:
```
--- FAIL: TestAccDataSourceMetalSpotMarketRequest_Basic (10.15s)
        	* metro: '' expected type 'string', got unconvertible type 'packngo.Metro'
```

Other failures were unrelated and do to availability (it would be nice to address this):
```
--- FAIL: TestAccMetalPortVlanAttachment_L2Individual (2.61s)
        Error: The facility nrt1 has no provisionable s1.large.x86 servers matching your criteria
 
--- FAIL: TestAccMetalPortVlanAttachment_HybridBasic (3.05s)
       Error: The facility dfw2 has no provisionable n2.xlarge.x86 servers matching your criteria

--- FAIL: TestAccMetalPortVlanAttachment_HybridMultipleVlans (2.64s) 
        Error: The facility nrt1 has no provisionable s1.large.x86 servers matching your criteria

--- FAIL: TestAccMetalPortVlanAttachment_L2Native (2.47s)
        Error: The facility nrt1 has no provisionable s1.large.x86 servers matching your criteria
```

**Comment by displague** [comment]
The SpotMarketRequest tests are passing.

Some additional (unrelated) failures were picked up:
```
--- FAIL: TestAccMetalPortVlanAttachment_HybridBasic (226.60s)
    testing.go:569: Step 0 error: Check failed: Check 1/1 error: metal_device.test: Attribute 'network_type' expected "layer3", got "hybrid"

--- FAIL: TestAccMetalPortVlanAttachment_HybridMultipleVlans (237.85s) (2021/05/19 20:32:57 - 20:33:03)
    testing.go:569: Step 1 error: errors during apply:
        Error: POST https://api.equinix.com/metal/v1/ports/d0dae5d1-8b47-4425-8514-f2e52ef99dc3/assign: 500 Oh snap, something went wrong! We've logged the error and will take a look - please reach out to us if you continue having trouble. 
          on /tmp/tf-test075083516/main.tf line 28:
          (source code not available)
        
        Error: POST https://api.equinix.com/metal/v1/ports/d0dae5d1-8b47-4425-8514-f2e52ef99dc3/assign: 500 Oh snap, something went wrong! We've logged the error and will take a look - please reach out to us if you continue having trouble. 
          on /tmp/tf-test075083516/main.tf line 28:
          (source code not available)
        
        Error: POST https://api.equinix.com/metal/v1/ports/d0dae5d1-8b47-4425-8514-f2e52ef99dc3/assign: 500 Oh snap, something went wrong! We've logged the error and will take a look - please reach out to us if you continue having trouble. 
          on /tmp/tf-test075083516/main.tf line 28:
          (source code not available)        
```

**Comment by t0mk** [comment]
.. taking a look

File: metal/resource_metal_spot_market_request.go [hunk 1]
```diff
@@ -359,6 +380,31 @@ func resourceMetalSpotMarketRequestDelete(d *schema.ResourceData, meta interface
...
+		"locked":           params.Locked,
+		"project_ssh_keys": params.ProjectSSHKeys,
+		"user_ssh_keys":    params.UserSSHKeys,
+		"userdata":         params.UserData,
+		"customdata":       params.CustomData,
+		"ipxe_script_url":  params.IPXEScriptURL,
+		"tags":             params.Tags,
+	}}
+	if params.TerminationTime != nil {
+		p[0]["termintation_time"] = params.TerminationTime.Time
```
Comment (t0mk) [inline metal/resource_metal_spot_market_request.go:403]:
```suggestion
		p[0]["termination_time"] = params.TerminationTime.Time
```

**Comment by t0mk** [comment]
@displague I see, you skipped filling the instance params whatsoever. Well, I guess it doesn't matter that much, the device_ids are most important there. Well, looks good enough.

**Review by t0mk — APPROVED** [review]

**Pull request merged**